push/sched/scheduler.go

```diff
@@ -2,29 +2,41 @@ package sched
 
 import (
 	log "github.com/sirupsen/logrus"
-	"sync"
+	"github.com/squareup/pranadb/cluster"
+	"github.com/squareup/pranadb/common"
+	"sync/atomic"
+	"time"
 )
 
+const maxProcessBatchRows = 500
+
 type ShardScheduler struct {
-	shardID uint64
-	actions chan *actionHolder
-	lock    sync.Mutex
-	started bool
-	paused  bool
+	shardID                  uint64
+	actions                  chan struct{}
+	started                  bool
+	failed                   bool
+	lock                     common.SpinLock
+	processorRunning         bool
+	forwardRows              []cluster.ForwardRow
+	batchHandler             RowsBatchHandler
+	lastQueuedReceiverSeq    uint64
+	lastProcessedReceiverSeq uint64
 }
 
-type Action func() error
+type RowsBatchHandler interface {
+	HandleBatch(shardID uint64, rows []cluster.ForwardRow) error
+}
 
-type actionHolder struct {
-	action  Action
-	errChan chan error
-	exit    bool
+type BatchEntry struct {
+	WriteTime time.Time
+	Seq       uint32
 }
 
-func NewShardScheduler(shardID uint64) *ShardScheduler {
+func NewShardScheduler(shardID uint64, batchHandler RowsBatchHandler) *ShardScheduler {
 	return &ShardScheduler{
-		shardID: shardID,
-		actions: make(chan *actionHolder, 1000), // TODO make configurable
+		shardID:      shardID,
+		actions:      make(chan struct{}, 10),
```

Comment:
This buffer size seems arbitrary. I think it's pretty rare for the buffer size to be different than 0 or 1. Do you mind adding a comment that explains the custom size?

Reply:
+1 It should work with channel size of 1